### ISSUE TITLE
Race condition when forcing a future

### ISSUE THREAD

**Issue by mdbergmann** [lmj/lparallel#38]
Hi.

I'm doing some benchmarking for my Actor based library 'cl-gserver'.
The whole thing is based on a 'mailbox' that is supposed to queue messages on one thread.

So I'm playing with 'mailboxes' with an lparallel kernel with 1 worker and
- submitting messages via 'submit-task/receive-result'
- submitting via 'future/force'

The submitting of messages is running on a lparallel kernel of 8 workers with something like this:
```
(defun runner-lp2 ()
  (setf *msgbox* (make-instance 'cl-gserver.messageb:message-box))
  (setf lparallel:*kernel* (lparallel:make-kernel 8))
  (setf *counter* 0)

  (unwind-protect
       (time
        (progn 
          (map nil #'lparallel:force
               (mapcar (lambda (x)
                         (lparallel:future
                           (dotimes (n 200000)
                             (msg-submit))))
                       (mapcar (lambda (n) (format nil "thread-~a" n))
                               (loop for n from 1 to 8 collect n))))
          (format t "Counter: ~a~%" *counter*)))
    (format t "Counter: ~a~%" *counter*)
    (lparallel:end-kernel)
    (cl-gserver.messageb:stop *msgbox*)))
```

Altogether there are 1.6m messages being sent evenly split across 8 threads towards the said 'mailbox' with one worker lparallel kernel.

The 'submit-task/receive-result' approach is a bit slower, but all messages are handled by the 'mailbox'.
The 'future/force' variant however does not process all 1.6m messages. It's always a bit less, like 157xyz, etc.

The 'mailbox' side looks like this:

```
(defmethod submit ((self message-box) message withreply-p handler-fun)
  (let* ((*kernel* (slot-value self 'message-kernel))
         (fut (future
                (funcall handler-fun message))))
    (with-slots (processed-messages) self
      (incf processed-messages))
    (if withreply-p
        (force fut)
        t)))
```

I don't understand why this happens.
I've tried a different driver purely using bt-threads to exclude that is something on the driving side.
I am expecting that 'force' blocks the same way as 'receive-result' does on the thread that does submit the message.

Can anyone tell me or give me some pointers?

Cheers,
Manfred

**Comment by mdbergmann** [comment]
Alright, I've stripped it down to just this. See the attachement.

And here the results from an sbcl repl:

```
* (runner-sr)
Evaluation took:
  1.943 seconds of real time
  12.444503 seconds of total run time (7.115314 user, 5.329189 system)
  [ Run times consist of 0.028 seconds GC time, and 12.417 seconds non-GC time. ]
  640.50% CPU
  6,201,663,223 processor cycles
  164,114,064 bytes consed

Counter: 1600000
NIL
* (runner-fut)
Evaluation took:
  0.451 seconds of real time
  3.426464 seconds of total run time (3.408803 user, 0.017661 system)
  [ Run times consist of 0.027 seconds GC time, and 3.400 seconds non-GC time. ]
  759.65% CPU
  1,438,189,204 processor cycles
  255,334,224 bytes consed

Counter: 1462452
NIL
```

As can be seen, the 'runner-fut' doesn't seem to process all the calls.

I'm, not sure if I'm doing something wrong, considering I'm new to CL.

[bench-stripped.txt](https://github.com/lmj/lparallel/files/4690642/bench-stripped.txt)

**Comment by anticrisis** [comment]
I don't know if this is a red herring, but the [`force-promise`](https://github.com/lmj/lparallel/blob/9c11f40018155a472c540b63684049acc9b36e15/src/promise.lisp#L129-L135) function's use of condition variables seems dubious to me, regardless of language. My understanding is that a condition variable must be guarded by a lock, which must be held before `condition-wait` (or `condition-notify`) is invoked. But this code doesn't use a lock.

That's certainly what the Bordeaux Threads manual [specifies](https://trac.common-lisp.net/bordeaux-threads/wiki/ApiDocumentation#Resourcecontention:conditionvariables). 

So I wouldn't trust this code. Maybe try modifying it to handle locks correctly? Should be pretty simple.

**Comment by mdbergmann** [comment]
Not sure about `force-promise`. Could be an issue.

Here however, `force-future` is called.
It seems that when `force` is called on a `future` and the `future` was not yet fulfilled, then it is taken out of the scheduler/kernel handling and body of the `future` is executed in another task-context.

I have identified three different scenarios. Two of then work OK, one doesn't.
1. OK - when all futures are fulfilled before force is called.
2. OK - when all futures are fulfilled by force.
3. NOK - when we have a mix of those, like when producer fills the queue faster than they are processed. Seems like a race condition of some kind.

**Comment by mdbergmann** [comment]
To conclude further.
The race condition seems to be between `force-future` and `with-unfulfilled-future/no-wait` where both call `fulfill-plan/call`. Somewhere this `canceledp` makes the trouble.
However, I'm not sure that manually calling `fulfill-plan/call` is the right thing to do on `force`.

**Comment by anticrisis** [comment]
None of the code in this project uses condition variables in the way I expect, and in the way the Bordeaux Threads manual specifies. Here's another example, which is implicated in your `runner-fut` test case: https://github.com/lmj/lparallel/blob/9c11f40018155a472c540b63684049acc9b36e15/src/cons-queue.lisp#L74-L80

(You can `TRACE` the function to see that it is called repeatedly in your case.)

The only time I've seen results like those in your test case is when locks are not being used correctly. I have to assume that's what's happening here. Perhaps the library's author can comment on why his use of condition variables without locks is correct?

I am not an expert in this area, but I've also mis-used condition variables in this way and seen race conditions exactly like this.

~~What I would do if I had time would be to patch the dozen or so places where condition variables are used in lparallel and wrap them in `with-lock-held` as we're supposed to do, and see if the race condition is resolved.~~ Be aware that both `condition-wait` and `condition-notify` have to hold the lock.

EDIT: our comments crossed, I wasn't ignoring your previous comment on purpose.

EDIT 2: `define-locking-function` is a macro which defines two functions, one of which holds a lock. But if I expand the macro, I still see generated code which uses condition variables without the lock being held. In fact, the annotation `no-lock` is appended to these function names. To my understanding, any code with no lock should not be using condition variables. Again, maybe the author can explain why this is ok, and your test case will be quite helpful.

**Comment by mdbergmann** [comment]
I'm not certain that a lock on the cons-queue is actually necessary.
There is code that can lock the cons-queue, but it seems unused.

If lparallel can make certain that only one worker makes access to the queue there is no need to lock it. So I believe the locking is happening on the workers.

The combo `submit-task/receive-result` is working fine.
It seems that the futures are using this core functionality as well, but `force future` is trying to 'steal' the task from the default scheduler.
But the futures do use a lock to lock the 'fulfillment'.
Still, something is not working right.

**Comment by anticrisis** [comment]
I added `(assert (sb-thread:holding-mutex-p lock))` right before `condition-notify` and `condition-wait` in `push-cons-queue` and `pop-cons-queue` and the assertions did not fire, indicating that the lock is being properly held in those cases when running your `runner-fut` test. (To be sure my assertions were compiled, I negated the assertion and it did fire.)

So it looks like condition variables is not the issue.

(By the way, I duplicate your issue setting `+per-thread+` to 200 rather than 200000, so even a fast test case demonstrates the problem, at least on my machine.)

**Comment by mdbergmann** [comment]
Thanks for looking into it.
Indeed, a lower bound to 'per-thread' will replicate the issue.

**Comment by anticrisis** [comment]
Ok, I had a theory about why your test was breaking, but my attempt to fix it uncovered another thing I couldn't explain, which may provide a clue.

You are ending your test run before all the workers have executed. It's confusing because I think the outside `(map nil #'lparallel:force ...)` will block until all work is done, but it only blocks until all the `submit-fun`s have been called. A call to `submit-msg-fut` returns immediately, but continues to execute in various threads.

However, my attempt to move the `force` outside of `submit-msg-fut` uncovered more behavior I couldn't explain. The main clue is this: if you *remove* the call to `force`, everything works correctly, but only if you wait for `end-kernel` to complete. Here is a working example:

```common-lisp
(defun submit-msg-fut-2 ()
  (let* ((lparallel:*kernel* *msgbox-kernel*)
         (fut (lparallel:future
                   (incf *counter*)
                   *counter*)))
    fut))

(defun runner-2 (submit-fun)
  (setf *msgbox-kernel* (lparallel:make-kernel 1))
  (let ((lparallel:*kernel* *msgbox-kernel*))
    (setf *msgbox-channel* (lparallel:make-channel)))
  (setf lparallel:*kernel* (lparallel:make-kernel +threads+))
  (setf *counter* 0)

  (unwind-protect
       (time
        (mapcar #'lparallel:force
                (mapcar (lambda (n)
                          (declare (ignore n))
                          (lparallel:future
                            (dotimes (n +per-thread+)
                              (funcall submit-fun))))
                        (loop for n from 1 to +threads+ collect n))))

    (format t "Counter before end: ~a~%" *counter*)
    (lparallel:end-kernel :wait t)
    (let ((lparallel:*kernel* *msgbox-kernel*))
      (lparallel:end-kernel :wait t))
    (format t "Final Counter: ~a~%" *counter*)))

(defun runner-fut-2 ()
  (runner-2 #'submit-msg-fut-2))
```

```common-lisp
;; running 8 x 10000

CL-USER> (runner-fut-2)

Evaluation took:
  0.032 seconds of real time
  0.281250 seconds of total run time (0.281250 user, 0.000000 system)
  878.13% CPU
  63,678,394 processor cycles
  13,187,888 bytes consed
  
Counter before end: 39599
Final Counter: 80000
(NIL NIL NIL NIL NIL NIL NIL NIL)
```

However, if you add a `force` back in `submit-msg-fut-2` or if you wrap the `funcall submit-fun` in a `force`, the test breaks. I have no explanation for this. So, this is broken code:

```common-lisp
(defun submit-msg-fut-2 ()
  (let* ((lparallel:*kernel* *msgbox-kernel*)
         (fut (lparallel:future
                   (incf *counter*)
                   *counter*)))
    (lparallel:force fut)))
```

Logically, we should be able to create futures in any number of threads, and then `force` them all, causing our calling thread to wait. But something is going on here where forcing a future causes unexpected results.

The next area to test would be to try to duplicate this behavior in an environment that doesn't use a separate `*msgbox-kernel*`, but I am moving on to other things today.

**Comment by mdbergmann** [comment]
I had a similar suspicion, that some tasks still run after the 'runner' ended.

So I also tried to check repeatedly for a few seconds after the `map nil (force...` ended.
But the counter did not increase anymore.
And `force` must block.

> The next area to test would be to try to duplicate this behavior in an environment that doesn't use a separate *msgbox-kernel*, but I am moving on to other things today.

Not sure. I had also tried a pure bordeaux-thread runner like this:

```
          (map nil #'bt:join-thread
               (mapcar (lambda (x)
                         (bt:make-thread (lambda ()
                                           (dotimes (n +per-thread+)
                                             (msg-submit)))
                                         :name x))
                       (mapcar (lambda (n) (format nil "thread-~a" n))
                               (loop for n from 1 to +threads+ collect n))))
```

To exclude that the problem is on the 'runner' side.

**Comment by mdbergmann** [comment]
I think that taking the future task out of the scheduler and fulfilling it manually is not such a good idea. It's tricky for quick to execute tasks, like those of this ticket.
I have also not seen this in other future implementations.
Instead 'force' should block until the scheduler has processed the future.

I have something working. Sometimes, on SBCL there is still a race condition where `+no-result+` is returned instead of the calculated result.
It's based on a condition variable where the calling thread waits until the result was generated.

```
(defslots plan (promise-base)
  ((fn :reader plan-fn :type (or null function))
   (cvar :initform (make-condition-variable))))

(defun force-future (future)
  ;; If we are here then we've stolen the task from the kernel.
  (with-%future-slots (canceledp cvar lock result) future
    (cond
      ((eq result +no-result+)
       (progn
         (when (eq result +no-result+)
           (condition-wait cvar lock))
         result))
      (t result))))

(defmacro with-unfulfilled-future/no-wait (future &body body)
  (with-gensyms (lock cvar canceledp result)
    `(with-%future-slots
         ((,lock lock) (,cvar cvar) (,canceledp canceledp) (,result result)) ,future
       (unless ,canceledp
         (with-lock-held (,lock)
           ,@body))
       (condition-notify ,cvar))))
```

**Comment by mdbergmann** [comment]
Some more thoughts:

I think the future implementations `force` feature is flawed.
`force` actually fulfills the future in the `force` call thread, while the `*kernel*` grabs a new task, another future maybe.
This should not happen at all.
In my case I explicitly want a one-threaded kernel in order to queue work.
'forcing' futures makes them execute in the 'force' caller thread instead of the single kernel worker.
Which means the futures are not queued anymore but are executed concurrently.

So I think the futures should be redesigned, or 'force' should be removed completely.

**Comment by anticrisis** [comment]
It still makes me nervous to see code using `condition-wait` and `condition-notify` without an enclosing `with-lock-held` anywhere. The condition variables will be subject to race conditions if they are not mutexed. Also, condition variables can be 'notified' spuriously, so they're not to be relied upon to definitively signal that a particular state has changed.

I've never seen a use of lock-free condition variables in other lock-free code I've reviewed. Usually other atomics are used if the code is specifically designed to be lock-free.

> ```common-lisp
> (defslots plan (promise-base)
>   ((fn :reader plan-fn :type (or null function))
>    (cvar :initform (make-condition-variable))))
> 
> (defun force-future (future)
>   ;; If we are here then we've stolen the task from the kernel.
>   (with-%future-slots (canceledp cvar lock result) future
>     (cond
>       ((eq result +no-result+)
>        (progn
>          (when (eq result +no-result+)
>            (condition-wait cvar lock))
>          result))
>       (t result))))
> 
> (defmacro with-unfulfilled-future/no-wait (future &body body)
>   (with-gensyms (lock cvar canceledp result)
>     `(with-%future-slots
>          ((,lock lock) (,cvar cvar) (,canceledp canceledp) (,result result)) ,future
>        (unless ,canceledp
>          (with-lock-held (,lock)
>            ,@body))
>        (condition-notify ,cvar))))
> ```

**Comment by mdbergmann** [comment]
Yeah. This code is a bit questionable.

I've just implemented something with lock and cvar with Bordeaux-threads, and usually you have to do it like so:

```
(withlock-held (lock)
  (condition-wait cvar lock)
  (do-something-else))
```

And on the other side:

```
(with-lock-held (lock)
  (do-something)
  (condition-notify cvar))
```

Other than that I couldn't get it working relyably.

I'll see if I can fix this with a lock and cvar combo and not steal it from the scheduler.
The fulfill function should be removed.

**Comment by anticrisis** [comment]
> ```common-lisp
> (with-lock-held (lock)
>   (condition-wait cvar lock)
>   (do-something-else))
> ```

I would revise as:

 ```common-lisp
(with-lock-held (lock)
  (condition-wait cvar lock)
  (when (state-was-actually-changed)
    (do-something-else)))
```

because `condition-wait` can return even if `condition-notify` was not called in the way the source code seems to expect. Here are some links on the 'spurious wakeup' issue, from C++:

* https://en.cppreference.com/w/cpp/thread/condition_variable
* https://en.wikipedia.org/wiki/Spurious_wakeup

**Comment by lmj** [comment]
Using `incf` on the same variable concurrently is a race condition, whether done via tasks or futures. The problem is resolved by using atomic operations such as SBCL's `sb-ext:atomic-incf`. You could also use the (unadvertised) `lparallel.counter` package which wraps `sb-ext:atomic-incf` etc. [Gist](https://gist.github.com/lmj/00ff903bd289ed5ec28913c48c2de034/revisions).

    CL-USER> (in-package #:bench-stripped)
    #<PACKAGE "BENCH-STRIPPED">
    BENCH-STRIPPED> (runner-sr)
    Evaluation took:
      5.068 seconds of real time
      24.268000 seconds of total run time (8.532000 user, 15.736000 system)
      [ Run times consist of 0.044 seconds GC time, and 24.224 seconds non-GC time. ]
      478.85% CPU
      17,192,628,797 processor cycles
      153,083,088 bytes consed
      
    Counter: 1600000
    NIL
    BENCH-STRIPPED> (runner-fut)
    Evaluation took:
      0.307 seconds of real time
      1.644000 seconds of total run time (1.616000 user, 0.028000 system)
      [ Run times consist of 0.008 seconds GC time, and 1.636 seconds non-GC time. ]
      535.50% CPU
      1,041,049,922 processor cycles
      255,406,272 bytes consed
      
    Counter: 1600000
    NIL

The `/no-lock` functions are for use inside `with-locked-queue`, in case you want to do more than one action while a queue is locked; see the [queue documentation](https://lparallel.org/api/queues/).

There are also internal (unexported) functions which are only called when the lock is held, like `force-future`.

Incidentally, I would be very careful about mixing CLOS with threads, even if particular SBCL bugs like [this one](https://bugs.launchpad.net/sbcl/+bug/1153309) appear to be (finally) fixed.

Is there anything left to be resolved here?

**Comment by mdbergmann** [comment]
> Using incf on the same variable concurrently is a race condition, whether done via tasks or futures.

Well. It is not changed concurrently because I chose a one worker kernel so that only one thread will have access to the global variable. This is effectively how Actors, or Clojures Agent work. State is hidden behind a single threaded queue.
As long as the future is not `force`d everything is fine. But as I've outlined above, `force`ing completes the future is another (the callers) thread which makes this whole construct concurrent.

> Is there anything left to be resolved here?

My opinion is that ' forcing' should not use the callers thread to fulfill the future.
It's your project. So you decide what to do with it.

**Comment by lmj** [comment]
OK, I should have looked more closely at why you were using a single-worker kernel.

The counter *is* changed concurrently---on purpose and by design---because of work stealing. If a future is unfulfilled, and you force it, then the future is stolen from the kernel and computed right there where you forced it. So the main thread is competing with the worker.

Without stealing, many auto-parallelized functions defined via `defpun` would not achieve speedup, among other slowdowns.

Futures, at least in the lparallel sense, don't seem to be a good match for what you're doing. I think the "Message passing" section [here](https://lparallel.org/kernel/) is closer to what you need. Make a single kernel with multiple workers and send messages through queues. @dimitri's pgloader uses this model with lparallel. Would that work for you?

**Comment by mdbergmann** [comment]
OK. If this is by design, then it is so. We can close this issue.

I was making some performance comparisons with submit-task/receive-result vs. futures for the library I'm working on [cl-gserver](https://github.com/mdbergmann/cl-gserver).
At the time I didn't know that futures are based on submit-task/receive-result and hence can't be faster.
During those performance tests I've realised this issue with concurrency on a single worker kernel.

I've also experimented with a pure single Bordeaux-threads thread for operating on the queue, which in my case is still 10% faster than a single worker kernel.
But you're cons-queue is really fast. I'm using that for an unbounded message queue.

**Comment by lmj** [comment]
I would add that it *has* to be by design, given lparallel's purpose of speeding up code. Consider

    (plet ((a (progn
                (sleep 1)
                (+ 1 2)))
           (b (progn
                (sleep 1)
                (+ 3 4))))
      (force (future (+ a b))))

Suppose there are only two workers. Execution order is generally indeterminate, so one worker may get `a` and the other may get `(+ a b)`. Without stealing, the total execution time would be about 2 seconds instead of 1 in this case. One worker would be blocked on `(+ a b)` while the other worker would execute the `plet` tasks in succession. With stealing, this needless waiting is avoided and the execution time will always be about 1 second, as expected.

I'm not sure a single-worker kernel makes sense in lparallel, especially if performance is the goal, as it turns the multi-threaded stealing scheduler and multi-threaded error handler into dead weight. I would have to think about this use case a little more, but it seems like using a boot as a glove---it works, in a way, but not what it was designed for.

Note if you don't need `pop-queue` to block, there is a lockless queue in the unadvertised package `lparallel.spin-queue`.

I am familiar with the experience of bug report responders not quite understanding my point. Sorry about that!

**Comment by lmj** [comment]
@anticrisis, have your concerns been resolved? Any problems that you still see?

**Comment by mdbergmann** [comment]
> I would add that it has to be by design, given lparallel's purpose of speeding up code. Consider

I think this should be made more clear how `force` actually works, that a thread is used outside of the kernel to force fulfilling it. Maybe in the promise documentation.

> With stealing, this needless waiting is avoided and the execution time will always be about 1 second, as expected.

That maybe true. However, it would be important to note, that the call thread of `force` potentially does the computation of one or more futures, instead of the workers of the kernel.
Usually one would expect that an explicitly defined  execution context (a thread-pool) is responsible for working on the futures. Otherwise it can be a source of subtle bugs.
Some applications maybe have multiple different defined execution contexts (thread-pools/*kernel*s). One for I/O that can have hundreds of workers and one for CPU bound computations where the *kernel* should match the number of workers to the available CPU cores.
The construct of `force` would add a third execution context that is somewhat undefined.

So while `force` may have its use-case there should/could be a function to wait for a future being fulfilled by the workers.

> but it seems like using a boot as a glove---it works, in a way, but not what it was designed for.

Yeah. I thought so, too. lparallel lives from having scenarios with > 1 workers.

**Comment by lmj** [comment]
I would add that it *has* to be by design. Consider

    (plet ((a (progn
                (sleep 1)
                (+ 1 2)))
           (b (progn
                (sleep 1)
                (+ 3 4))))
      (force (future (+ a b))))

Suppose there are only two workers. Task execution order is generally indeterminate, so one worker may get `b` and the other may get `(+ a b)`. Without stealing, the total execution time would be about 2 seconds instead of 1 in this case. One worker would be blocked on `(+ a b)` while the other worker would execute the `plet` tasks in succession. With stealing, this needless waiting is avoided and the execution time will be about 1 second, as expected.

I'm not sure a single-worker kernel makes sense in lparallel, especially if performance is the goal, as it turns the multi-threaded stealing scheduler and multi-threaded error handler into dead weight. I would have to think about this use case a little more, but it seems like using a boot as a glove---it works, in a way, but not what it was designed for.

Note if there are places where `pop-queue` doesn't need to block, there is a lockless queue in the unadvertised package `lparallel.spin-queue`.

I am familiar with the experience of bug report responders not quite understanding my point. Sorry about that!

**Comment by lmj** [comment]
I would add that it *has* to be by design, given lparallel's purpose of speeding up code. Consider

    (plet ((a (progn
                (sleep 1)
                (+ 1 2)))
           (b (progn
                (sleep 1)
                (+ 3 4))))
      (force (future (+ a b))))

Suppose there are only two workers. Execution order is generally indeterminate, so one worker may get `b` and the other may get `(+ a b)`. Without stealing, the total execution time would be about 2 seconds instead of 1 in this case. One worker would be blocked on `(+ a b)` while the other worker would execute the `plet` tasks in succession. With stealing, this needless waiting is avoided and the execution time will always be about 1 second, as expected.

I'm not sure a single-worker kernel makes sense in lparallel, especially if performance is the goal, as it turns the multi-threaded stealing scheduler and multi-threaded error handler into dead weight. I would have to think about this use case a little more, but it seems like using a boot as a glove---it works, in a way, but not what it was designed for.

Note if there are places where `pop-queue` doesn't need to block, there is a lockless queue in the unadvertised package `lparallel.spin-queue`.

I am familiar with the experience of bug report responders not quite understanding my point. Sorry about that!

**Comment by anticrisis** [comment]
> @anticrisis, have your concerns been resolved? Any problems that you still see?

Thank you for the ping, I think the discussion you've had has been quite helpful in clarifying the design intent of lparallel futures. I agree with @mdbergmann that it seems unexpected. I was going to suggest updating the documentation of `force`, but it seems you've already thought of that:

```
    Note if `force' is called on an unfulfilled future then the future is
    fulfilled by the caller of `force'.
```

Many thanks.

**Comment by lmj** [comment]
That note in the documentation of `force` could be more prominent, and it should begin "Importantly, if" instead of "Note if".

To summarize, there are two very different models of parallelism in play. In one model, the aim is to achieve speedup by keeping all processors busy via work stealing, like in

    (defpun fib (n)
      (if (< n 2)
          n
          (plet ((a (fib (- n 1)))
                 (b (fib (- n 2))))
            (+ a b))))

Here the auto-parallelizing engine breaks up the code into tasks, with task sizes being determined dynamically as processors become free.

In the other model, the user has already broken up code into tasks and has assigned each task to a specific processor. While this isn't lparallel's native model, a simple abstraction on top of lparallel can give it to you.

Here is what I came up with: [wires](https://gist.github.com/lmj/08657a9bcea572b479712094c1a557aa). Each wire is tied directly to an individual worker.

    (defun test-names ()
      (with-wires (wire-a wire-b)
        (send wire-a (lambda () (+ 1 2)))
        (send wire-b (lambda () (+ 3 4)))
        (assert (= 3 (receive wire-a)))
        (assert (= 7 (receive wire-b)))))
    
    (defun test-future-override ()
      (with-wires (wire-a wire-b)
        (with-wired-future
          (let ((f (future wire-a (+ 1 2)))
                (g (future wire-b (+ 3 4))))
            (assert (= 3 (force f)))
            (assert (= 7 (force g)))))))

This should be about as fast as raw bordeaux-threads since lparallel's machinery is mainly in the setup. Note if `task-handler-bind` is used then it needs to be outside of `with-wires`, which is where the lparallel tasks are created.

**Comment by mdbergmann** [comment]
So this `with-wired-future` will `force` all futures within the available *kernel* workers?

**Comment by lmj** [comment]
I don't understand the question. All workers are kernel workers, at least in lparallel terminology. `with-wires` takes over the kernel and attaches the wires. One may then `send` and `receive` along those wires. `future` and `force` are locally-defined wrappers around `send` and `receive`, respectively. For simplicity it is assumed there are not competing threads calling `force` on the same future (the native `force` in lparallel does not have this limitation).

**Comment by mdbergmann** [comment]
What I meant is, `force` 'steals' from the workers if the future is not fulfilled. In that case the future is _not_ fulfilled by the workers.
Does `force` when called inside `with-wired-future` also steal, or is the future fulfilled by the `workers` and `force` blocks until that happened?

**Comment by lmj** [comment]
`force` doesn't steal inside `with-wired-future`, which has a local redefinition of `force` which calls `receive` which calls `pop-queue` on the wire. If the future has already been fulfilled, `force` returns the cached value.

**Comment by mdbergmann** [comment]
Hi.
It's been a while. The 'wires', could you put this into a release?
